### PR TITLE
Storing timing data per endpoint instance

### PR DESCRIPTION
**Pull request by tmasternak**
Connects to https://github.com/Particular/ServiceControl/issues/979

### Goal
Store information about number of physical instances for given logical endpoint.

### PoA
 - [x] Parse `instanceId` passed from `NServiceBus.Metrics`
 - [x] Store information about known physical instances 
 - [x] Expose number of physical instances in data returned form API 
 - [x] Update ATT to newest version of `NServiceBus.Metrics`

/cc: @WilliamBZA

### CODE REVIEW

File: src/ServiceControl.Monitoring/EndpointInstanceId.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+            var hostId = headers[Headers.OriginatingHostId];
+            var hostDisplayName = headers[Headers.HostDisplayName];
+
+            return new EndpointInstanceId(endpointName, hostId, hostDisplayName);
+        }
+
+        protected bool Equals(EndpointInstanceId other)
+        {
+            return string.Equals(EndpointName, other.EndpointName) &&
```
Comment (WilliamBZA) [inline src/ServiceControl.Monitoring/EndpointInstanceId.cs]: Is it not enough here to simply check the Instance Id?

If 2 instances have the same endpoint name, the only difference will be the instance Id. If they have a different endpoint name, the instance Id will also be different.

File: src/ServiceControl.Monitoring/EndpointInstanceId.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+            return Equals((EndpointInstanceId) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = (EndpointName != null ? EndpointName.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (InstanceId != null ? InstanceId.GetHashCode() : 0);
```
Comment (WilliamBZA) [inline src/ServiceControl.Monitoring/EndpointInstanceId.cs]: Same as the equals comparer, can we not simply use the instance Id here?

File: src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        /// <summary>
+        /// Initializes the metric API module.
+        /// </summary>
+        public MonitoredEndpointsModule(ProcessingTimeStore processingTimeStore, CriticalTimeStore criticalTimeStore)
+        {
+            Get["/monitored-endpoints"] = x =>
+            {
+                var endpointsData = TimingsAggregator.AggregateIntoLogicalEndpoints(processingTimeStore, criticalTimeStore);
+
+                return Negotiate.WithModel(endpointsData.ToArray());
```
Comment (WilliamBZA) [inline src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs]: Don't need the `ToArray()` here.

File: src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ServiceControl.Monitoring.Timings
+{
+    class TimingsAggregator
+    {
+        public static List<MonitoredEndpoint> AggregateIntoLogicalEndpoints(ProcessingTimeStore processingTimeStore, CriticalTimeStore criticalTimeStore)
```
Comment (WilliamBZA) [inline src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs]: I wonder if this code couldn't look a bit better by changing the return type to `IEnumerable` and then `yield return`ing the MonitoredEndpoint instance instead of adding to a full list

File: src/ServiceControl.Monitoring/EndpointInstanceId.cs (hunk 1 above)
Comment (mikeminutillo) [inline src/ServiceControl.Monitoring/EndpointInstanceId.cs]: I think it should be possible to host 2 endpoints in the same host. That can lead to a situation where two endpoints have different names but the same instance id.

File: src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            var now = DateTime.UtcNow;
+
+            var processingTimes = processingTimeStore.GetTimings(now);
+            var criticalTimes = criticalTimeStore.GetTimings(now);
+
+            var endpointNames = new[] { processingTimes, criticalTimes }
+                .SelectMany(t => t).Select(t => t.Id.EndpointName)
```
Comment (mikeminutillo) [inline src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs]: Can the `SelectMany` and `Select` be collapsed into a single `SelectMany`? Either that or a `processingTimes.Concat(criticalTimes).Select(t => t.Id.EndpointName).Distinct()` could do the job.

File: src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs [hunk 6]
```diff
@@ -16,7 +16,7 @@ public QueueLengthDataStore(IQueueLengthCalculator calculator)
             get { return calculator.GetQueueLengths().Select(kvp => new KeyValuePair<string, JObject>(kvp.Key, new JObject(new JProperty("Count", kvp.Value)))); }
         }
 
-        public void Store(JObject data)
+        public void Store(string endpointName, string instanceId, JObject data)
```
Comment (mikeminutillo) [inline src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs]: Why is this needed? These values don't get used.

File: src/ServiceControl.Monitoring/EndpointInstanceId.cs (hunk 1 above)
Comment (tmasternak) [inline src/ServiceControl.Monitoring/EndpointInstanceId.cs]: > I think it should be possible to host 2 endpoints in the same host. That can lead to a situation where two endpoints have different names but the same instance id.

That's a good point. Did not realize that 👍

File: src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs (hunk 4 above)
Comment (tmasternak) [inline src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs]: 👍 readability for the way.

File: src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs (hunk 6 above)
Comment (tmasternak) [inline src/ServiceControl.Monitoring/QueueLength/QueueLengthDataStore.cs]: Removed

File: src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs (hunk 5 above)
Comment (tmasternak) [inline src/ServiceControl.Monitoring/Http/Diagrams/TimingsAggregator.cs]: I like the `Concat` it reads a lot better. 👍

**Comment by tmasternak** [comment]
@mikeminutillo I've addressed review comments

**Comment by tmasternak** [comment]
Can we merge this one?
